salt/postgres/files/postgres-backup-to-s3.sh

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+#
+# Backup PostgreSQL databases and upload to AWS S3
+
+set -euo pipefail
+
+PSQL_ADMIN_USER="postgres"
+
+# shellcheck disable=SC1091
+. /home/sysadmin-tools/aws-settings.local
+
+export AWS_ACCESS_KEY_ID
+export AWS_SECRET_ACCESS_KEY
+export AWS_DEFAULT_REGION
+
+if [ "$LOGNAME" != "postgres" ]; then
+    echo "ERROR: Execution of $0 stopped as not run by user postgres!"
+    exit 2
+fi
+
+if [ ! -x "$AWS_CLI" ]; then
+    echo "Error: The aws executable is not installed"
+    exit 3
+fi
+
+# Create work directory resolving permissions alerts when running psql from root.
+WORKDIR=$(mktemp -d)
+chown $PSQL_ADMIN_USER:$PSQL_ADMIN_USER "$WORKDIR"
+
+mapfile -t DATABASES < <(cd "$WORKDIR" && sudo -u $PSQL_ADMIN_USER /usr/bin/psql -t --csv -c 'select datname from pg_database')
+
+# Using read over mapfile because the latter leaves a newline on the final item when processing a space delimited string.
+read -ra REQUESTED_DATABASES <<< "${BACKUP_DATABASES}"
+
+for DATABASE in "${REQUESTED_DATABASES[@]}"; do
+    if [[ "${DATABASES[*]}" =~ $DATABASE ]]; then
```

Comment:
Instead of a condition, why not let `pg_dump` fail? I suppose the condition prevents `mktemp` from creating files that will not be cleaned up, but I think it's OK to create some unused temp files.

Reply:
Because we have `set -euo pipefail` if a backup fails it would stop the script from continuing preventing other database backups. Handling this situation lets us send a clear error message and the backups can continue.

I have not tested how `pg_dump` acts when given a malformed database name, I expect it returns a non-zero exit code. I hope `pg_dump` also returned a clear error message - which we would need if we were to remove this check.

Reply:
Aha, good call.

FYI, `pg_dump foo` returns status code 1 and:

```
pg_dump: error: connection to database "foo" failed: FATAL:  database "foo" does not exist
```